docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+# Sync improvements
+
+* Author(s): Cornelius Weig (@corneliusweig)
+* Design Shepherd: Tejal Desai (@tejal29)
+* Date: 03/20/2019
+* Status: Draft
+
+## Background
+
+Currently, skaffold config supports `artifact.sync` as a way to sync files
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807).
+
+Furthermore, builders should be able to inform Skaffold about additional sync paths (#1704).
+This will result in much faster deploy cycles, without users having to bother about these optimizations.
+
+## Design
+
+### Goals
+The sync mechanism in skaffold should be able to infer the destination location of local files.
+
+For example: given the following Dockerfile
+
+```dockerfile
+FROM nginx
+WORKDIR /var
+COPY static *.html www/
+ADD nginx.conf .
+ADD nginx.conf /etc/
+```
+
+The sync locations should be inferred as
+- `static/page1.html` -> `/var/www/page1.html` (! no static folder at destination !)
+- `static/assets/style.css` -> `/var/www/assets/style.css`
+- `index.html` -> `/var/www/index.html`
+- `nginx.conf` -> `/var/nginx.conf`, `/etc/nginx.conf`
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Whether jib and bazel builders can provide that information too, is unclear at the moment.
```

Comment:
Jib can provide this information.  For Jib, the static files are usually in `src/main/resources/` and then copied into the container image under `/app/resources`.  So Jib could return a sync map of `src/main/resources/** -> /app/resources`.

Reply:
Just to clarify: is `src/main/resources` below the workingdir of the artifact? If this is not the case, we need to think how we can handle absolute paths on the host as well, because the current assumption is that `SyncMap` returns _relative_ paths wrt the workingdir.

Reply:
Hmm… it should be.  Jib assumes the artifact context points to a Maven/Gradle project.  If it's a standalone project, it should have `src/main/...`.  If it's a multi-module project, where there are several subproject — like libraries — that are joined together to produce an image, there will be multiple `xxx/src/main/...` directories, but they should all be under the context directory too.